File: 12_Local_Database_Connection_with_Docker/04_MySQL_Connection_with_Python/connect_database.py
# pip install mysql-connector-python

import os
import sys
import mysql.connector
from mysql.connector import Error

# connection config (use env vars or defaults)
MYSQL_HOST = os.environ.get("MYSQL_HOST", "host.docker.internal")
# MYSQL_HOST = os.environ.get("MYSQL_HOST", "localhost")  # uncomment for non-Docker setup
MYSQL_PORT = int(os.environ.get("MYSQL_PORT", 3306))
MYSQL_USER = os.environ.get("MYSQL_USER", "root")
MYSQL_PASSWORD = os.environ.get("MYSQL_PASSWORD", "")
MYSQL_DB = os.environ.get("MYSQL_DB", "e-commerce")  # DB shown in your phpMyAdmin URL

def connect_mysql(host: str, port: int, user: str, password: str, database: str):
    try:
        conn = mysql.connector.connect(
            host=host,
            port=port,
            user=user,
            password=password,
            database=database,
            connection_timeout=5,
        )
        if conn.is_connected():
            print(f"✅ Connected to MySQL DB '{database}' on {host}:{port}")
            return conn
        raise Error("Connection not established")
    except Exception as e:
        print("❌ MySQL connection error:", e, file=sys.stderr)
        raise

def fetch_products(conn, database: str = MYSQL_DB, table: str = "products"):
    cursor = None
    try:
        cursor = conn.cursor(dictionary=True)
        # Quote identifiers to allow hyphens in database/table names and avoid injection
        db_quoted = f"`{database.replace('`','')}`"
        table_quoted = f"`{table.replace('`','')}`"
        sql = f"SELECT * FROM {db_quoted}.{table_quoted};"
        cursor.execute(sql)
        rows = cursor.fetchall()
        print("📄 Products from DB:")
        for r in rows:
            print(r)
    except Exception as e:
        print("❌ Error fetching products:", e, file=sys.stderr)
    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()

if __name__ == "__main__":
    conn = connect_mysql(MYSQL_HOST, MYSQL_PORT, MYSQL_USER, MYSQL_PASSWORD, MYSQL_DB)
    fetch_products(conn)

File: 12_Local_Database_Connection_with_Docker/02_MySQL_Connection_with_NodeJS/connect_database.js
import mysql from 'mysql2/promise';
import dotenv from 'dotenv';

dotenv.config();

// Connection config (use env vars or defaults)
const MYSQL_HOST = process.env.MYSQL_HOST || 'host.docker.internal';
// const MYSQL_HOST = process.env.MYSQL_HOST || 'localhost'; // uncomment for non-Docker setup
const MYSQL_PORT = parseInt(process.env.MYSQL_PORT || '3306');
const MYSQL_USER = process.env.MYSQL_USER || 'root';
const MYSQL_PASSWORD = process.env.MYSQL_PASSWORD || '';
const MYSQL_DB = process.env.MYSQL_DB || 'e-commerce'; // DB shown in your phpMyAdmin URL

async function connectMysql(host, port, user, password, database) {
  try {
    const conn = await mysql.createConnection({
      host,
      port,
      user,
      password,
      database,
      waitForConnections: true,
      connectionLimit: 1,
      queueLimit: 0,
    });
    console.log(`✅ Connected to MySQL DB '${database}' on ${host}:${port}`);
    return conn;
  } catch (e) {
    console.error('❌ MySQL connection error:', e.message);
    throw e;
  }
}

async function fetchProducts(conn, database = MYSQL_DB, table = 'products') {
  try {
    // Backticks escape identifiers to allow hyphens in database/table names and avoid injection
    const dbQuoted = `\`${database.replace(/`/g, '')}\``;
    const tableQuoted = `\`${table.replace(/`/g, '')}\``;
    const sql = `SELECT * FROM ${dbQuoted}.${tableQuoted};`;
    
    const [rows] = await conn.execute(sql);
    console.log('📄 Products from DB:');
    rows.forEach(r => console.log(r));
  } catch (e) {
    console.error('❌ Error fetching products:', e.message);
  } finally {
    if (conn) {
      await conn.end();
    }
  }
}

(async () => {
  const conn = await connectMysql(MYSQL_HOST, MYSQL_PORT, MYSQL_USER, MYSQL_PASSWORD, MYSQL_DB);
  await fetchProducts(conn);
})();